### PR TITLE
Catkin plugin: Add support for ROS tools.

### PR DESCRIPTION
**Pull request by kyrofa**
The current Catkin plugin distributes only the built version of a ROS package that would typically go into the devel workspace, but that version doesn't include non-source, non-binary files, such as .launch files. This PR changes the plugin to actually install ROS packages, thereby including any non-source, non-binary files specified by the developer.

This PR also updates the wrapper environment to correctly setup the ROS environment, which allows for the use of typical ROS tools within the snapcraft.yaml, such as roslaunch, rosrun, etc., and updates the ROS examples to make use of them.

Finally, this PR also increases the test coverage of the Catkin plugin twofold.

### CODE REVIEW

File: examples/ros/src/beginner_tutorials/CMakeLists.txt [hunk 1]
```diff
@@ -34,6 +32,9 @@ add_executable(add_two_ints_client src/add_two_ints_client.cpp)
...
-# %EndTag(SRVCLIENT)%
-
-# %EndTag(FULLTEXT)%
\ No newline at end of file
+## Mark executables for installation
+install(TARGETS talker listener add_two_ints_server add_two_ints_client
+  ARCHIVE DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
+  LIBRARY DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
+  RUNTIME DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION}
+)
```
Comment (sergiusens) [inline examples/ros/src/beginner_tutorials/CMakeLists.txt]: :100:

File: snapcraft/plugins/catkin.py [hunk 2]
```diff
@@ -105,6 +105,7 @@ def env(self, root):
                 root,
                 self.options.rosversion),
             'ROS_MASTER_URI=http://localhost:11311',
```
Comment (sergiusens) [inline snapcraft/plugins/catkin.py]: Not part of the PR, just something to keep in mind.

File: snapcraft/tests/test_plugin_catkin.py [hunk 3]
```diff
@@ -29,16 +30,67 @@ class _IOError(IOError):
...
+                      <package>
+                        <buildtool_depend>buildtool_depend</buildtool_depend>
+                        <build_depend>build_depend</build_depend>
+                        <run_depend>run_depend</run_depend>
+                      </package>"""
+
+        plugin = catkin.CatkinPlugin('test-part', self.properties)
+
+        with mock.patch.object(builtins, 'open',
+                               mock.mock_open(read_data=package_xml)):
```
Comment (elopio) [inline snapcraft/tests/test_plugin_catkin.py]: Could you remove this mock by writting the xml file to the sourcedir?

File: snapcraft/plugins/catkin.py (hunk 2 above)
Comment (kyrofa) [inline snapcraft/plugins/catkin.py]: Heh, I don't quite understand what you're trying to draw my attention to. That you'd like that variable removed eventually?

File: snapcraft/tests/test_plugin_catkin.py [hunk 4]
```diff
@@ -29,16 +30,67 @@ class _IOError(IOError):
...
+            plugin.pull()
+
+        self.ubuntu_mock.assert_has_calls([
+            mock.call().get([
+                'ros-foo-buildtool-depend',
+                'ros-foo-build-depend',
+                'ros-foo-run-depend']),
+            mock.call().unpack(plugin.installdir)])
```
Comment (elopio) [inline snapcraft/tests/test_plugin_catkin.py]: This looks good.

File: snapcraft/plugins/catkin.py [hunk 5]
```diff
@@ -105,6 +105,7 @@ def env(self, root):
...
+            'ROS_HOME=$SNAP_APP_USER_DATA_PATH/.ros',
```
Comment (sergiusens) [inline snapcraft/plugins/catkin.py]: To avoid future misgivings, can we add a comment on why this is needed?

File: snapcraft/tests/test_plugin_catkin.py (hunk 3 above)
Comment (kyrofa) [inline snapcraft/tests/test_plugin_catkin.py]: Indeed I could, but I feel like actually hitting the filesystem when I don't need to feels a bit brittle and slow.

File: snapcraft/plugins/catkin.py [hunk 6]
```diff
@@ -245,12 +246,18 @@ def build(self):
 
         self.run(
             ['rm', '-f',
-             'opt/ros/' + self.options.rosversion + '/.catkin',
-             'opt/ros/' + self.options.rosversion + '/.rosinstall',
-             'opt/ros/' + self.options.rosversion + '/setup.sh',
-             'opt/ros/' + self.options.rosversion + '/_setup_util.py'],
+             'opt/ros/' + self.options.rosversion + '/.rosinstall'],
```
Comment (sergiusens) [inline snapcraft/plugins/catkin.py]:
```python
os.remove(os.path.join(self.installdir, self.options.rosversion, '.rosinstall))
```

File: snapcraft/plugins/catkin.py [hunk 7]
```diff
@@ -245,12 +246,18 @@ def build(self):
...
             cwd=self.installdir)
 
+        # Fix the shebang in _setup_util.py to be portable
```
Comment (elopio) [inline snapcraft/plugins/catkin.py]: It would be nice to have a unittest for the shebang.
Aren't we doing this already in the python plugin? Maybe we can share some code, and unit test that.

File: snapcraft/tests/test_plugin_catkin.py (hunk 3 above)
Comment (elopio) [inline snapcraft/tests/test_plugin_catkin.py]: In unit tests, I prefer to access the filesystem in temp directories instead of mocking that part.
But your test is correct, if you prefer it this way, I'm ok with that.

File: snapcraft/plugins/catkin.py (hunk 5 above)
Comment (kyrofa) [inline snapcraft/plugins/catkin.py]: Good call, done.

File: snapcraft/plugins/catkin.py (hunk 2 above)
Comment (sergiusens) [inline snapcraft/plugins/catkin.py]: That maybe it eventually is expected to run on a different port. But I'm not sure that is configurable.

File: snapcraft/plugins/catkin.py (hunk 6 above)
Comment (kyrofa) [inline snapcraft/plugins/catkin.py]: Haha, alright I'll clean that up now.

File: snapcraft/plugins/catkin.py (hunk 7 above)
Comment (sergiusens) [inline snapcraft/plugins/catkin.py]: It's been moved to meta. If we want to share it I'd move it to another module.

File: snapcraft/plugins/catkin.py (hunk 6 above)
Comment (elopio) [inline snapcraft/plugins/catkin.py]: This deserves an integration test, to check that the files you want end up in the right dirs.
Feel free to add that in a later branch, because that's debt from before.

File: snapcraft/plugins/catkin.py (hunk 7 above)
Comment (kyrofa) [inline snapcraft/plugins/catkin.py]: I agree, but ALL the build steps here need test coverage, which I plan on adding in a later PR.

File: snapcraft/plugins/catkin.py (hunk 6 above)
Comment (sergiusens) [inline snapcraft/plugins/catkin.py]: Oh, and if you know why it is being removed, a comment would be great :smile:

File: snapcraft/plugins/catkin.py (hunk 2 above)
Comment (kyrofa) [inline snapcraft/plugins/catkin.py]: Ah, okay. It is actually configurable, I just need to figure out how to do that with Snappy, which I'll make a different PR.

File: snapcraft/tests/test_plugin_catkin.py (hunk 3 above)
Comment (kyrofa) [inline snapcraft/tests/test_plugin_catkin.py]: I don't feel strongly about this. I'll update it :) .

**Comment by sergiusens** [comment]
I like this very much. One followup question is; do we need the `ros` plugin at all? I always found it redundant as maybe the core can be pulled in through some logic here.

**Comment by kyrofa** [comment]
@sergiusens I was wondering that myself. Let me experiment, this PR may have enabled the removal of it.

File: snapcraft/plugins/catkin.py (hunk 6 above)
Comment (kyrofa) [inline snapcraft/plugins/catkin.py]: @elopio agreed 100%. I'll do it in a later PR though.

Comment (ted-gould) [inline snapcraft/plugins/catkin.py]: It shouldn't need to be removed, but can be filtered now that the filters are fixed. Never got around to replacing the rm with proper stage filters.

**Comment by kyrofa** [comment]
@sergiusens very close. As soon as https://bugs.launchpad.net/snappy/+bug/1466234 is resolved we can remove the roscore plugin.

File: snapcraft/plugins/catkin.py (hunk 6 above)
Comment (kyrofa) [inline snapcraft/plugins/catkin.py]: I got rid of all the removals here, none of them seemed to be necessary.

File: snapcraft/tests/test_plugin_catkin.py (hunk 3 above)
Comment (kyrofa) [inline snapcraft/tests/test_plugin_catkin.py]: Done.

**Comment by kyrofa** [comment]
Alright I believe I've resolved all the concerns for this one. I will continue to increase test coverage throughout subsequent PRs, and will remove the roscore plugin once the [blocking bug](https://bugs.launchpad.net/snappy/+bug/1466234) has been resolved.

**Comment by elopio** [comment]
The example fails to build with:
Error: parts catkin-tutorials and roscore have the following file paths in common which have different contents:

  usr/bin/xml2-config

**Comment by kyrofa** [comment]
@elopio oh interesting, I didn't consider the interplay between plugins! I will add that to my manual test procedure. xml2-config isn't a required file, so I have temporarily added the removal of it back to the catkin plugin until we can remove roscore.

**Comment by sergiusens** [comment]
@elopio and @kyrofa fwiw, plugins aren't supposed to solve all the conflicts in the world, that is to some extent part of the part author's responsibility. That does no rule out plugins themselves should try to deliver the cleanest environment possible, but not in detriment of its functionality.

In the example of `xml-config`, if I as an author ever create another part that also ships `xml-config` I would need to solve if somehow, luckily, that is somewhat what the `stage`, `snap` and `organize` keywords are for.

**Comment by sergiusens** [comment]
I'm :+1: on this branch btw :wink:

**Comment by elopio** [comment]
Me too. Merging.
